File: 算法/代码随想录算法PDF/00调试源代码/438.找到字符串中所有字母异位词.py
#
# @lc app=leetcode.cn id=438 lang=python3
#
# [438] 找到字符串中所有字母异位词
#

# @lc code=start
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        m , n = len(s), len(p)
        res = []
        if m < n: return res
        smap = [0] * 26
        pmap = [0] * 26
        for i in range(n):
            smap[ord(s[i]) - ord('a')] += 1
            pmap[ord(p[i]) - ord('a')] += 1
        if smap == pmap: res.append(0)
        for i in range(n, m):
            smap[ord(s[i]) - ord('a')] += 1
            smap[ord(s[i - n]) - ord('a')] -= 1
            if smap == pmap: res.append(i - n + 1)
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/438.找到字符串中所有字母异位词.cpp
/*
 * @lc app=leetcode.cn id=438 lang=cpp
 *
 * [438] 找到字符串中所有字母异位词
 */

// @lc code=start
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int m = s.length(), n = p.length();
        vector <int> res;
        if (n > m) {  // 特殊情况
            return res;
        }

        vector <int> smap(26, 0), pmap(26, 0);  // 字典
        for (int i = 0; i < n; i++) {
            smap[s[i] - 'a']++;
            pmap[p[i] - 'a']++;
        }
        if (smap == pmap) {
            res.push_back(0);
        }
        for (int i = n; i < m; i++) {
            smap[s[i] - 'a']++;
            smap[s[i-n] - 'a']--;
            if (smap == pmap) {
                res.push_back(i - n + 1);
            }
        }
        return res;
    }
};
// @lc code=end

